packages/lodestar/src/sync/unknownBlock.ts

```diff
@@ -0,0 +1,295 @@
+    private readonly logger: ILogger,
+    private readonly metrics: IMetrics | null,
+    opts?: SyncOptions
+  ) {
+    if (!opts?.disableUnknownBlockSync) {
+      this.network.events.on(NetworkEvent.unknownBlockParent, this.onUnknownBlock);
+      this.network.events.on(NetworkEvent.peerConnected, this.triggerUnknownBlockSearch);
+    }
+
+    if (metrics) {
+      metrics.syncUnknownBlock.pendingBlocks.addCollect(() =>
+        metrics.syncUnknownBlock.pendingBlocks.set(this.pendingBlocks.size)
+      );
+      metrics.syncUnknownBlock.knownBadBlocks.addCollect(() =>
+        metrics.syncUnknownBlock.knownBadBlocks.set(this.knownBadBlocks.size)
+      );
+    }
+  }
+
+  close(): void {
+    this.network.events.off(NetworkEvent.unknownBlockParent, this.onUnknownBlock);
+    this.network.events.off(NetworkEvent.peerConnected, this.triggerUnknownBlockSearch);
+  }
+
+  /**
+   * Process an unknownBlockParent event and register the block in `pendingBlocks` Map.
+   */
+  private onUnknownBlock = (signedBlock: allForks.SignedBeaconBlock, peerIdStr: string): void => {
+    try {
+      this.addToPendingBlocks(signedBlock, peerIdStr);
+      this.triggerUnknownBlockSearch();
+      this.metrics?.syncUnknownBlock.requests.inc(1);
+    } catch (e) {
+      this.logger.error("Error handling unknownBlockParent event", {}, e as Error);
+    }
+  };
+
+  private addToPendingBlocks(signedBlock: allForks.SignedBeaconBlock, peerIdStr: string): PendingBlock {
+    const block = signedBlock.message;
+    const blockRoot = this.config.getForkTypes(block.slot).BeaconBlock.hashTreeRoot(block);
+    const blockRootHex = toHexString(blockRoot);
+    const parentBlockRootHex = toHexString(block.parentRoot);
+
+    let pendingBlock = this.pendingBlocks.get(blockRootHex);
+    if (!pendingBlock) {
+      pendingBlock = {
+        blockRootHex,
+        parentBlockRootHex,
+        signedBlock,
+        peerIdStrs: new Set(),
+        status: PendingBlockStatus.pending,
+        downloadAttempts: 0,
+      };
+      this.pendingBlocks.set(blockRootHex, pendingBlock);
+    }
+    pendingBlock.peerIdStrs.add(peerIdStr);
+
+    // Limit pending blocks to prevent DOS attacks that cause OOM
+    const prunedItemCount = pruneSetToMax(this.pendingBlocks, MAX_PENDING_BLOCKS);
+    if (prunedItemCount > 0) {
+      this.logger.warn(`Pruned ${prunedItemCount} pending blocks from UnknownBlockSync`);
+    }
+
+    return pendingBlock;
+  }
+
+  /**
+   * Gather tip parent blocks with unknown parent and do a search for all of them
+   */
+  private triggerUnknownBlockSearch = (): void => {
+    const blocks = this.pendingBlocks.size > 0 && getLowestPendingUnknownParents(this.pendingBlocks);
+
+    // Cheap early stop to prevent calling the network.getConnectedPeers()
+    if (!blocks) {
+      return;
+    }
+
+    // If the node loses all peers with pending unknown blocks, the sync will stall
+    const connectedPeers = this.network.getConnectedPeers();
```

Comment:
is it better to only get peers with `status.head >= ours`? it's no use to ask for blocks from syncing peers

Reply:
Oh good point! Will do in another PR :pray: